### PR TITLE
[Feature] Add MonoFlex Coder

### PR DESCRIPTION
**Pull request by ZCMax**
Thanks for your contribution and we appreciate it a lot. The following instructions would make your pull request more healthy and more easily get feedback. If you do not understand some items, don't worry, just make the pull request and seek help from maintainers.

## Motivation
Add monoflex coder.

## Modification
add monoflex_bbox_coder.py

## BC-breaking (Optional)
Does the modification introduce changes that break the back-compatibility of the downstream repos?
If so, please describe how it breaks the compatibility and how the downstream projects should modify their code to keep compatibility with this PR.

## Use cases (Optional)
If this PR introduces a new feature, it is better to list some use cases here, and update the documentation.

## Checklist

1. Pre-commit or other linting tools are used to fix the potential lint issues.
2. The modification is covered by complete unit tests. If not, please add more unit test to ensure the correctness.
3. If the modification has potential influence on downstream projects, this PR should be tested with downstream projects.
4. The documentation has been modified accordingly, like docstring or example tutorials.

### CODE REVIEW

File: mmdet3d/core/bbox/coders/monoflex_bbox_coder.py [hunk 1]
```diff
@@ -0,0 +1,387 @@
...
+    Args:
+        depth_mode (str): The depth mode for direct depth calculation.
+        base_depth (tuple[float]): Depth references for decode box depth.
+        depth_range (str): Depth range of predicted depth.
+        use_combined_depth (bool): Whether to use combined depth (direct depth
+            and depth from keypoints).
+        dims_mean (tuple[tuple[float]]): Dimensions mean of decode bbox
+            dimensions [l, h, w] for each category.
+        dims_mean (tuple[tuple[float]]): Dimensions std of decode bbox
+            dimensions [l, h, w] for each category.
```
Comment (Tai-Wang) [inline mmdet3d/core/bbox/coders/monoflex_bbox_coder.py:22]: Merge mean and std as a tuple[tuple[float]] like FCOS3DBBoxCoder and name it as base_dims? Just extend the definition of base_depth

File: mmdet3d/core/bbox/coders/monoflex_bbox_coder.py [hunk 2]
```diff
@@ -0,0 +1,387 @@
...
+        dims_modes (list[str|bool]): Dimensions modes.
```
Comment (Tai-Wang) [inline mmdet3d/core/bbox/coders/monoflex_bbox_coder.py:23]: Specify mode choices

File: mmdet3d/core/bbox/coders/monoflex_bbox_coder.py [hunk 3]
```diff
@@ -0,0 +1,387 @@
...
+        # output related
+        self.bbox_code_size = code_size
+        self.eps = 1e-3
+
+    def encode(self, locations, dimensions, orientations, input_metas):
+        """Encode CameraInstance3DBoxes by locations, dimensions, orientations.
+
+        Args:
+            locations (Tensor): Center location for 3D boxes.
+                (N, 3)
```
Comment (Tai-Wang) [inline mmdet3d/core/bbox/coders/monoflex_bbox_coder.py:73]: Center location for 3D boxes of shape (N,3).
Please also unify the format of the following doc of arguments.

File: mmdet3d/core/bbox/coders/monoflex_bbox_coder.py [hunk 4]
```diff
@@ -0,0 +1,387 @@
...
+
+        Args:
+            reg (Tensor): Batch regression for each predict center2d point.
+                shape: (batch * K (max_objs), C)
+            points(Tensor): Batch projected bbox centers on image plane.
+                shape: (batch * K (max_objs) , 2)
+            labels (Tensor): Batch predict class label for each predict
+                center2d point.
+                shape: (batch * K (max_objs), )
+            down_ratio (int): The stride of feature map.
```
Comment (Tai-Wang) [inline mmdet3d/core/bbox/coders/monoflex_bbox_coder.py:105]: downsample_ratio

File: mmdet3d/core/bbox/coders/monoflex_bbox_coder.py [hunk 5]
```diff
@@ -0,0 +1,387 @@
...
+
+        pred_offsets2d = reg[:, 4:6]
+        pred_dimensions_offsets3d = reg[:, 29:32]
+        pred_orientations = torch.cat((reg[:, 32:40], reg[:, 40:48]), dim=1)
+        # decode the pred residual dimensions to real dimensions
+        # (B* max_objs, 3)
+        pred_dimensions = self.decode_dimension(labels,
+                                                pred_dimensions_offsets3d)
+        # decode the depth with direct regression
+        pred_depth_offsets = reg[:, 48:49].squeeze(-1)
```
Comment (Tai-Wang) [inline mmdet3d/core/bbox/coders/monoflex_bbox_coder.py:128]: Need some comments to specify which dimensions correspond to which predictions at the beginning

File: mmdet3d/core/bbox/coders/monoflex_bbox_coder.py [hunk 6]
```diff
@@ -0,0 +1,387 @@
...
+                - orientations (Tensor): Orientations of 3D
+                    boxes.
+                    shape: (batch * K (max_objs), 1)
+        """
+
+        pred_offsets2d = reg[:, 4:6]
+        pred_dimensions_offsets3d = reg[:, 29:32]
+        pred_orientations = torch.cat((reg[:, 32:40], reg[:, 40:48]), dim=1)
+        # decode the pred residual dimensions to real dimensions
+        # (B* max_objs, 3)
```
Comment (Tai-Wang) [inline mmdet3d/core/bbox/coders/monoflex_bbox_coder.py:124]: B * max_objs

File: mmdet3d/core/bbox/coders/monoflex_bbox_coder.py [hunk 7]
```diff
@@ -0,0 +1,387 @@
...
+                        centers2d_target,
+                        offsets2d,
+                        depths,
+                        cam2imgs,
+                        down_ratio,
+                        pad_mode='default'):
+        """Retrieve object location.
+
+        Args:
+            centers2d_target (torch.Tensor): Projected 3D target centers
```
Comment (Tai-Wang) [inline mmdet3d/core/bbox/coders/monoflex_bbox_coder.py:204]: From my understanding, does it mean base_centers2d? We can get the predicted centers2d from base_centers2d and offsets2d? centers2d_target can be a little confusing.

File: mmdet3d/core/bbox/coders/monoflex_bbox_coder.py (hunk 5 above)
Comment (ZCMax) [reply mmdet3d/core/bbox/coders/monoflex_bbox_coder.py:128]: I think we can refactor the code by adding another class: key2channels,  by using dimension str (like 'offsets2d')  to index the regression.

File: mmdet3d/core/bbox/coders/monoflex_bbox_coder.py [hunk 8]
```diff
@@ -0,0 +1,387 @@
...
+        batch_id = torch.arange(N_batch).unsqueeze(1)
+        obj_id = batch_id.repeat(1, N // N_batch).flatten()
+        # (B * max_objs, 4, 4)
+        cam2imgs = cam2imgs[obj_id]
+        # (B * max_objs, )
+        f_u = cam2imgs[:, 0, 0]
+        # (B * max_objs, )
+        center_height = pred_keypoints2d[:, -2, 1] - pred_keypoints2d[:, -1, 1]
+        # (B * max_objs, 2)
+        corner_02_height = \
```
Comment (Tai-Wang) [inline mmdet3d/core/bbox/coders/monoflex_bbox_coder.py:279]: Add a basic comment to explain what are 02 and 13

File: mmdet3d/core/bbox/coders/monoflex_bbox_coder.py [hunk 9]
```diff
@@ -0,0 +1,387 @@
...
+        else:
+            dimensions = dims_offset * cls_dimension_mean
+
+        return dimensions
+
+    def decode_orientation(self, vector_ori, locations):
+        """Retrieve object orientation.
+
+        Args:
+            vector_ori(torch.Tensor): Local vector orientation
```
Comment (Tai-Wang) [inline mmdet3d/core/bbox/coders/monoflex_bbox_coder.py:335]: add a space between 'vector_ori' and '(torch.Tensor)' and 'locations' and '(torch.Tensor)'

File: mmdet3d/core/bbox/coders/monoflex_bbox_coder.py [hunk 10]
```diff
@@ -0,0 +1,387 @@
...
+        """
+        if self.multibin:
+            pred_bin_cls = vector_ori[:, :self.num_dir_bin * 2].view(
+                -1, self.num_dir_bin, 2)
+            pred_bin_cls = torch.softmax(pred_bin_cls, dim=2)[..., 1]
+            orientations = vector_ori.new_zeros(vector_ori.shape[0])
+            for i in range(self.num_dir_bin):
+                mask_i = (pred_bin_cls.argmax(dim=1) == i)
+                s = self.num_dir_bin * 2 + i * 2
+                e = s + 2
```
Comment (Tai-Wang) [inline mmdet3d/core/bbox/coders/monoflex_bbox_coder.py:352]: start_bin
end_bin?

File: mmdet3d/core/bbox/coders/monoflex_bbox_coder.py [hunk 11]
```diff
@@ -0,0 +1,441 @@
...
+        batch_bboxes = input_metas[0]['box_type_3d'](
+            bboxes, box_dim=self.bbox_code_size)
+
+        return batch_bboxes
+
+    def decode(self, reg, labels, downsample_ratio, cam2imgs):
+        """Decode regression into 3D predictions.
+
+        Args:
+            reg (Tensor): Batch regression for each predict center2d point.
```
Comment (Tai-Wang) [inline mmdet3d/core/bbox/coders/monoflex_bbox_coder.py:98]: reg -> bbox (to be consistent with other bbox coders) / pred

**Review by Tai-Wang — COMMENTED** [review]
Almost ready to merge

File: mmdet3d/core/bbox/coders/monoflex_bbox_coder.py [hunk 12]
```diff
@@ -0,0 +1,453 @@
...
+    Args:
+        depth_mode (str): The depth mode for direct depth calculation.
+        base_depth (tuple[float]): Depth references for decode box depth.
+        depth_range (list): Depth range of predicted depth.
+        use_combined_depth (bool): Whether to use combined depth (direct depth
+            and depth from keypoints).
+        uncertainty_range (list): Uncertainty range of predicted depth.
+        base_dims (tuple[tuple[float]]): Dimensions mean and std of decode bbox
+            dimensions [l, h, w] for each category.
+        dims_modes (list[str|bool]): Dimensions modes. It should includes three
```
Comment (ZwwWayne) [inline mmdet3d/core/bbox/coders/monoflex_bbox_coder.py:22]: This description is not clear, may give some example.

File: mmdet3d/core/bbox/coders/monoflex_bbox_coder.py [hunk 13]
```diff
@@ -0,0 +1,453 @@
...
+
+        # orientation related
+        self.multibin = multibin
+        self.bin_centers = bin_centers
+        self.num_dir_bins = num_dir_bins
+        self.bin_margin = bin_margin
+
+        # output related
+        self.bbox_code_size = code_size
+        self.eps = 1e-3
```
Comment (ZwwWayne) [inline mmdet3d/core/bbox/coders/monoflex_bbox_coder.py:69]: make it settable

File: mmdet3d/core/bbox/coders/monoflex_bbox_coder.py [hunk 14]
```diff
@@ -0,0 +1,453 @@
...
+        base_depth (tuple[float]): Depth references for decode box depth.
+        depth_range (list): Depth range of predicted depth.
+        use_combined_depth (bool): Whether to use combined depth (direct depth
+            and depth from keypoints).
+        uncertainty_range (list): Uncertainty range of predicted depth.
+        base_dims (tuple[tuple[float]]): Dimensions mean and std of decode bbox
+            dimensions [l, h, w] for each category.
+        dims_modes (list[str|bool]): Dimensions modes. It should includes three
+            parts, [linear, log or exp ; use mean or not ; use std or not]
+        multibin (bool): Whether to use multi_bin representation.
```
Comment (ZwwWayne) [inline mmdet3d/core/bbox/coders/monoflex_bbox_coder.py:24]: multi_bin or multibin?

File: mmdet3d/core/bbox/coders/monoflex_bbox_coder.py [hunk 15]
```diff
@@ -0,0 +1,453 @@
...
+        depth_mode,
+        base_depth,
+        depth_range,
+        use_combined_depth,
+        uncertainty_range,
+        base_dims,
+        dims_modes,
+        multibin,
+        bin_centers,
+        num_dir_bins,
```
Comment (ZwwWayne) [inline mmdet3d/core/bbox/coders/monoflex_bbox_coder.py:44]: put bin_centers/bin_marge together.

File: mmdet3d/core/bbox/coders/monoflex_bbox_coder.py [hunk 16]
```diff
@@ -0,0 +1,453 @@
...
+                shape: (N, C)
+            labels (Tensor): Batch predict class label for each predict
+                center2d point.
+                shape: (N, )
+            downsample_ratio (int): The stride of feature map.
+            cam2imgs (Tensor): Batch images' camera intrinsic matrix.
+                shape: kitti (N, 4, 4)  nuscenes (N, 3, 3)
+
+        Return:
+            dict: The 3D prediction dict decoded from regression map.
```
Comment (ZwwWayne) [inline mmdet3d/core/bbox/coders/monoflex_bbox_coder.py:120]: indicate its keys.

File: mmdet3d/core/bbox/coders/monoflex_bbox_coder.py [hunk 17]
```diff
@@ -0,0 +1,453 @@
...
+            torch.cat(centers2d_img, depths.unsqueeze(-1), dim=1)
+        # (N, 4, 1)
+        centers2d_extend = \
+            torch.cat((centers2d_img, centers2d_img.new_ones(N, 1)),
+                      dim=1).unqueeze(-1)
+        locations = torch.matmul(cam2imgs_inv, centers2d_extend).squeeze(-1)
+
+        return locations[:, :3]
+
+    def decode_depth_from_keypoints(self,
```
Comment (ZwwWayne) [inline mmdet3d/core/bbox/coders/monoflex_bbox_coder.py:268]: decode_depth_from_keypoints -> keypoints2depth

File: mmdet3d/core/bbox/coders/monoflex_bbox_coder.py [hunk 18]
```diff
@@ -0,0 +1,453 @@
...
+        corner_group1_depth = corner_group1_depth.mean(dim=1)
+
+        keypoints_depth = torch.stack(
+            (center_depth, corner_group0_depth, corner_group1_depth), dim=1)
+        keypoints_depth = torch.clamp(
+            keypoints_depth, min=self.depth_range[0], max=self.depth_range[1])
+
+        return keypoints_depth
+
+    def decode_dimension(self, labels, dims_offset):
```
Comment (ZwwWayne) [inline mmdet3d/core/bbox/coders/monoflex_bbox_coder.py:368]: decode_dims

File: mmdet3d/core/bbox/coders/monoflex_bbox_coder.py [hunk 19]
```diff
@@ -0,0 +1,453 @@
...
+            direct_depth=pred_direct_depth,
+            keypoints2d=pred_keypoints2d,
+            keypoints_depth=pred_keypoints_depth,
+            combined_depth=pred_combined_depth,
+            orientations=pred_orientations,
+        )
+
+        return preds
+
+    def decode_direct_depth(self, depth_offsets):
```
Comment (ZwwWayne) [inline mmdet3d/core/bbox/coders/monoflex_bbox_coder.py:196]: what is the difference between direct depth and depth?

Comment (ZCMax) [reply mmdet3d/core/bbox/coders/monoflex_bbox_coder.py:211]: in monoflex depth should includes directly regressed depth and keypoints depth.

File: mmdet3d/core/bbox/coders/monoflex_bbox_coder.py [hunk 20]
```diff
@@ -0,0 +1,514 @@
...
+
+        # output related
+        self.bbox_code_size = code_size
+        self.eps = eps
+
+    def encode(self, gt_bboxes_3d):
+        """Encode ground truth to prediction targets.
+
+        Args:
+            gt_bboxes_3d (BaseInstance3DBoxes): Ground truth 3D bboxes.
```
Comment (Tai-Wang) [inline mmdet3d/core/bbox/coders/monoflex_bbox_coder.py:77]: \`BaseInstance3DBoxes\`

**Review by Tai-Wang — APPROVED** [review]

File: mmdet3d/core/bbox/coders/monoflex_bbox_coder.py [hunk 21]
```diff
@@ -0,0 +1,516 @@
...
+            for i in range(self.num_dir_bins):
+                mask_i = (pred_bin_cls.argmax(dim=1) == i)
+                start_bin = self.num_dir_bins * 2 + i * 2
+                end_bin = start_bin + 2
+                pred_bin_offset = ori_vector[mask_i, start_bin:end_bin]
+                orientations[mask_i] = torch.atan2(
+                    pred_bin_offset[:, 0],
+                    pred_bin_offset[:, 1]) + self.bin_centers[i]
+        else:
+            axis_cls = torch.softmax(ori_vector[:, :2], dim=1)
```
Comment (ZwwWayne) [inline mmdet3d/core/bbox/coders/monoflex_bbox_coder.py:440]: use tensor.softmax()

File: mmdet3d/core/bbox/coders/monoflex_bbox_coder.py [hunk 22]
```diff
@@ -0,0 +1,516 @@
...
+            axis_cls = axis_cls[:, 0] < axis_cls[:, 1]
+            head_cls = torch.softmax(ori_vector[:, 2:4], dim=1)
+            head_cls = head_cls[:, 0] < head_cls[:, 1]
+            # cls axis
+            orientations = self.bin_centers[axis_cls + head_cls * 2]
+            sin_cos_offset = F.normalize(ori_vector[:, 4:])
```
Comment (ZwwWayne) [inline mmdet3d/core/bbox/coders/monoflex_bbox_coder.py:446]: tensor.normalize()

File: mmdet3d/core/bbox/coders/monoflex_bbox_coder.py [hunk 23]
```diff
@@ -0,0 +1,516 @@
...
+            orientations += torch.atan(sin_cos_offset[:, 0] /
+                                       sin_cos_offset[:, 1])
+
+        locations = locations.view(-1, 3)
+        rays = torch.atan2(locations[:, 0], locations[:, 2])
```
Comment (ZwwWayne) [inline mmdet3d/core/bbox/coders/monoflex_bbox_coder.py:451]: tensor.atan2?

**Comment by ZwwWayne** [comment]
final review, can be merged after fix.

File: mmdet3d/core/bbox/coders/monoflex_bbox_coder.py (hunk 22 above)
Comment (ZCMax) [reply mmdet3d/core/bbox/coders/monoflex_bbox_coder.py:445]: It cannot be written as tensor.normalize() format.
